### PR TITLE
Feature/react save data twitter

### PR DESCRIPTION
**Pull request by anton-karlovskiy**
Not yet ready to be merged.

TODO:
1. react-window implementation -- done
2. react hook approach confirming
3. refactor requirement confirming
4. README writing -- done
5. Google's license -- done

### CODE REVIEW

**Comment by anton-karlovskiy** [comment]
@addyosmani 
In README of the whole repository there are two types of demos.

### Adaptive Loading patterns
### Full applications

Where shall we add this project?

**Comment by addyosmani** [comment]
I think this belongs in full applications.

**Comment by addyosmani** [comment]
Could you summarize how far we were able to get to with supporting Lite Mode triggering of the data saver mode on Android? 💡

File: react-twitter-save-data-loading/README.md [hunk 1]
```diff
@@ -1,3 +1,20 @@
...
-* [Link to Project on Glitch](https://glitch.com/~anton-karlovskiy-react-save-data-twitter/)
\ No newline at end of file
+
+# Demo: React-twitter save-data resource loading in React
+
+[React Tweet](https://github.com/mannynotfound/react-tweet) is an demo application looking like Twitter website.
+
+This version of React twitter uses Save-Data Client Hint to determine whether to load low-fidelity or high-fidelity images based on the user's Save-Data option on Android Chrome (e.g Lite Mode On -> lo-fi, Lite Mode Off -> hi-fi).
+
+[Live Demo](https://adaptive-loading.web.app/react-twitter-save-data-loading/)
```
Comment (addyosmani) [inline react-twitter-save-data-loading/README.md]: I just tested https://adaptive-loading.web.app/react-twitter-save-data-loading/ using Lite Mode on Android and it's working really great!

File: react-twitter-save-data-loading/README.md [hunk 2]
```diff
@@ -1,3 +1,20 @@
-## Glitch Source
-* [Link to Glitch App](https://anton-karlovskiy-react-save-data-twitter.glitch.me/)
-* [Link to Project on Glitch](https://glitch.com/~anton-karlovskiy-react-save-data-twitter/)
\ No newline at end of file
+
+# Demo: React-twitter save-data resource loading in React
+
+[React Tweet](https://github.com/mannynotfound/react-tweet) is an demo application looking like Twitter website.
+
+This version of React twitter uses Save-Data Client Hint to determine whether to load low-fidelity or high-fidelity images based on the user's Save-Data option on Android Chrome (e.g Lite Mode On -> lo-fi, Lite Mode Off -> hi-fi).
```
Comment (addyosmani) [inline react-twitter-save-data-loading/README.md]: `This is a demo of the Twitter timeline implemented using React and the Save-Data Client Hint. It determines whether to load low-fidelity or high-fidelity images based on the user's Data Saver (Lite Mode) preferences on Chrome for Android`.

**Comment by anton-karlovskiy** [comment]
> Could you summarize how far we were able to get to with supporting Lite Mode triggering of the data saver mode on Android? 💡

I'm sorry. To be honest I think I have not understood your question correctly. Could you please explain it again?

On Android Chrome, once we set Lite Mode to On in Settings, then every request includes `Save-Data` header by default.
So then we just call `Save-Data` API to the server. Actually that API does not have any special payload.
On the server side, just checking `Save-Data` request header from the client side we can make a decision whether to send down `light` or `heavy` signal.

On the client side, based on the signal from the server side, we can request `heavy` or `light` images.

File: react-twitter-save-data-loading/README.md (hunk 2 above)
Comment (anton-karlovskiy) [reply react-twitter-save-data-loading/README.md]: Thank you for writing README for me.

**Comment by anton-karlovskiy** [comment]
> I think this belongs in patterns under a Data Saver-style entry.

Yes, I updated the README so please review it.

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Please check the messages on slack channel as I left some messages for this project.

**Comment by anton-karlovskiy** [comment]
@addyosmani
This project is ready to be merged once you review the messages on slack channel and codebase.

**Comment by addyosmani** [comment]
Hmm. Approximately 3 days ago the deploy we had on Glitch was working correctly when Lite Mode was turned on. I just retested and for some reason it only now works when trying the manual client side mode from the app. Could we check what is deployed in Glitch is the latest?

**Comment by anton-karlovskiy** [comment]
> Hmm. Approximately 3 days ago the deploy we had on Glitch was working correctly when Lite Mode was turned on. I just retested and for some reason it only now works when trying the manual client side mode from the app. Could we check what is deployed in Glitch is the latest?

Yes, I'll check it. Could you please try firebase deployment? It's the latest.

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Yes, I found the reason.
First, the Glitch is only for manual client side mode as we don't have Node.js server there.
Just firebase functions can be working for it.
As well, for now, the save-data API is disappeared as I tried to deploy react-animation functions which does not include save-data API.
I deployed firebase functions again. Now it's working on https://adaptive-loading.web.app/react-twitter-save-data-loading/

**Comment by addyosmani** [comment]
Perfect! I can confirm it's working as expected again.

**Comment by addyosmani** [comment]
Screenshot_20191015-065904

Minor nit: can we fix the additional gap between tweet cards seen here on mobile?

It appears to not be present when viewed on desktop :)

Screenshot from desktop mode:

Screenshot_20191015-070050

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Yes, I have seen this. :)
I tried to fix it for a couple of hours but not reached the goal and the gap is made because we use [react-window](https://github.com/bvaughn/react-window).
Here with react-window we are supposed to set itemSize as static constant and looks like it's not adjusted automatically according to the device screen.
Let me try again.

**Comment by anton-karlovskiy** [comment]
@addyosmani 
Please try the [current version](https://adaptive-loading.web.app/react-twitter-save-data-loading/).
I tried to adjust the gap according to the device type (mobile or desktop) but I cannot but admit that this is not a perfect resolution. However I thought it might be ok for the purpose of demo.

**Comment by addyosmani** [comment]
This works :) Thanks, @anton-karlovskiy!

File: react-twitter-save-data-loading/src/App.js [hunk 3]
```diff
@@ -26,6 +26,13 @@ import './App.css';
...
+  return check;
+};
+
```
Comment (addyosmani) [inline react-twitter-save-data-loading/src/App.js]: I would like for us to circle back on this at some point as mobile detection based on UA strings rather than feature detection is generally not that reliable. Let's consider it OK for the demo purposes for now.

**Comment by anton-karlovskiy** [comment]
> This works :) Thanks, @anton-karlovskiy!

Thank you.

File: react-twitter-save-data-loading/src/App.js (hunk 3 above)
Comment (anton-karlovskiy) [reply react-twitter-save-data-loading/src/App.js]: OK, sounds good.
Let me log this note.
